cas/vaultcas/vaultcas.go

```diff
@@ -0,0 +1,322 @@
+			vc.RoleID,
+			&auth.SecretID{FromString: vc.SecretID},
+			auth.WithWrappingToken(),
+			auth.WithMountPath(vc.AppRole),
+		)
+	} else {
+		appRoleAuth, err = auth.NewAppRoleAuth(
+			vc.RoleID,
+			&auth.SecretID{FromString: vc.SecretID},
+			auth.WithMountPath(vc.AppRole),
+		)
+	}
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to initialize AppRole auth method")
+	}
+
+	authInfo, err := client.Auth().Login(ctx, appRoleAuth)
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to login to AppRole auth method")
+	}
+	if authInfo == nil {
+		return nil, errors.New("no auth info was returned after login")
+	}
+
+	return &VaultCAS{
+		client:      client,
+		config:      vc,
+		fingerprint: opts.CertificateAuthorityFingerprint,
+	}, nil
+}
+
+// CreateCertificate signs a new certificate using Hashicorp Vault.
+func (v *VaultCAS) CreateCertificate(req *apiv1.CreateCertificateRequest) (*apiv1.CreateCertificateResponse, error) {
+	switch {
+	case req.CSR == nil:
+		return nil, errors.New("CreateCertificate: `CSR` cannot be nil")
+	case req.Lifetime == 0:
+		return nil, errors.New("CreateCertificate: `LIFETIME` cannot be 0")
+	}
+
+	cert, chain, err := v.createCertificate(req.CSR, req.Lifetime)
+	if err != nil {
+		return nil, err
+	}
+
+	return &apiv1.CreateCertificateResponse{
+		Certificate:      cert,
+		CertificateChain: chain,
+	}, nil
+}
+
+func (v *VaultCAS) GetCertificateAuthority(req *apiv1.GetCertificateAuthorityRequest) (*apiv1.GetCertificateAuthorityResponse, error) {
+	secret, err := v.client.Logical().Read(v.config.PKI + "/cert/ca")
+	if err != nil {
+		return nil, errors.Wrap(err, "GetCertificateAuthority: unable to read root")
+	}
+	if secret == nil {
+		return nil, errors.New("GetCertificateAuthority: secret root is empty")
+	}
+
+	var certBundle certutil.CertBundle
+
+	err = mapstructure.Decode(secret.Data, &certBundle)
+	if err != nil {
+		return nil, err
+	}
+
+	cert, _, err := getCertificateAndChain(certBundle)
+	if err != nil {
+		return nil, err
+	}
+	return &apiv1.GetCertificateAuthorityResponse{
+		RootCertificate: cert,
+	}, nil
+}
+
+// RenewCertificate will always return a non-implemented error as renewals
+// are not supported yet.
+func (v *VaultCAS) RenewCertificate(req *apiv1.RenewCertificateRequest) (*apiv1.RenewCertificateResponse, error) {
+	return nil, apiv1.ErrNotImplemented{Message: "vaultCAS does not support renewals"}
+}
+
+func (v *VaultCAS) RevokeCertificate(req *apiv1.RevokeCertificateRequest) (*apiv1.RevokeCertificateResponse, error) {
+	if req.SerialNumber == "" && req.Certificate == nil {
+		return nil, errors.New("RevokeCertificate `serialNumber` or `certificate` are required")
+	}
+
+	serialNumber := req.SerialNumber
+	if req.Certificate != nil {
+		serialNumber = req.Certificate.SerialNumber.String()
+	}
+
+	serialNumberDash := strings.ReplaceAll(serialNumber, ":", "-")
+
+	_, err := v.client.Logical().Write(v.config.PKI+"/revoke/"+serialNumberDash, nil)
```

Comment:
It looks like Vault expects the serial number in hexadecimal format, you will get it as a big decimal number, you can use the big.Int to convert it to bytes and then encode it as vault expects.

Reply:
Serial numbers are "dashed", see https://github.com/hashicorp/vault/blob/56e81b899bdb3ce8f3bc28a36988f6a7cd37826b/builtin/logical/pki/path_revoke.go#L20